cartographer/mapping_3d/sparse_pose_graph.cc

```diff
@@ -393,35 +393,42 @@ void SparsePoseGraph::FreezeTrajectory(const int trajectory_id) {
   });
 }
 
-void SparsePoseGraph::AddSubmapFromProto(const int trajectory_id,
-                                         const transform::Rigid3d& initial_pose,
+void SparsePoseGraph::AddSubmapFromProto(const transform::Rigid3d& pose,
                                          const mapping::proto::Submap& submap) {
   if (!submap.has_submap_3d()) {
     return;
   }
 
+  const mapping::SubmapId proto_submap_id = {submap.submap_id().trajectory_id(),
+                                             submap.submap_id().submap_index()};
+  const int trajectory_id = proto_submap_id.trajectory_id;
   std::shared_ptr<const Submap> submap_ptr =
       std::make_shared<const Submap>(submap.submap_3d());
 
   common::MutexLocker locker(&mutex_);
   AddTrajectoryIfNeeded(trajectory_id);
   const mapping::SubmapId submap_id =
       submap_data_.Append(trajectory_id, SubmapData());
+  // TODO: remove this check when the optimization problem is refactored to
```

Comment:
Why is this different in 2D?

Reply:
The optimization problem still keeps its trajectory nodes in nested vector, so I could not write a `InsertTrajectoryNode` method for the optimization problem (which inserts a node with an exact ID from the proto) without going in the wrong direction. The ID remapping has been removed, so this means that `AddSubmapFromProto` is actually currently broken for trimmed data until `AddTrajectoryNode` is replaced with `InsertTrajectoryNode`. So I decided to keep everything as-is in 3D for the time being.

This is why I mentioned in #549 that it would be better to first merge MapByID for the SPG3D optimization problem. Then, I could make the change for 3D as I did with 2D.

Reply:
Where do you still see nested vectors? This should have been fixed by @cschuet and me in the last days.

Reply:
Sorry, I accidentally misread this when reading the header of the 3D optimization problem. My eyes slipped to another line and saw a vector of something, so I just assumed it hasn't been refactored yet. I will update this to exactly follow 2D.